src/interfaces/controllingiunknown.hpp

```diff
@@ -0,0 +1,77 @@
+#ifndef _SRC_INTERFACES_CONTROLLINGIUNKNOWN_HPP_
+#define _SRC_INTERFACES_CONTROLLINGIUNKNOWN_HPP_
+
+#include "tearoffbase.hpp"
+#include <dncp.h>
+#include <atomic>
+#include <vector>
+#include <new>
+#include <utility>
+
+class ControllingIUnknown final : public IUnknown
+{
+private:
+    std::atomic<int32_t> _refCount{ 1 };
+    std::vector<TearOffUnknown*> _tearOffs;
+public:
+    ControllingIUnknown() = default;
+
+    template<typename T, typename... Ts>
+    bool CreateAndAddTearOff(Ts... args)
+    {
+        T* tearOff = new (std::nothrow) T(this, std::forward<Ts>(args)...);
+        if (tearOff == nullptr)
+            return false;
+        _tearOffs.push_back(tearOff);
```

Comment:
Sadly we need to wrap this in a `try`/`catch` for the `std::bad_alloc const&` case.

Reply:
Right, `push_back` can fail 😢